h/security/policy/top_level.py

```diff
@@ -13,75 +10,34 @@
 from h.security.policy.helpers import is_api_request
 
 
-@implementer(ISecurityPolicy)
 class TopLevelPolicy(IdentityBasedPolicy):
-    """
-    The security policy for `h`.
-
-    This delegates to various different policies depending on the situation.
-    """
+    """The top-level security policy. Delegates to subpolicies."""
 
-    def __init__(self, proxy_auth=False):
-        """
-        Initialise a security policy.
-
-        :param proxy_auth: Replace the default `CookiePolicy` for the UI with
-            the `RemoteUserPolicy`.
-        """
-        self._bearer_token_policy = BearerTokenPolicy()
-        self._http_basic_auth_policy = AuthClientPolicy()
+    def __init__(self):
         self._identity_cache = RequestLocalCache(self._load_identity)
 
-        self._ui_policy = RemoteUserPolicy() if proxy_auth else CookiePolicy()
-
-    def remember(self, request, userid, **kw):
-        """Get the correct headers to remember the given user."""
-
+    def forget(self, request, **kw):
         self._identity_cache.clear(request)
+        return get_subpolicy(request).forget(request, **kw)
 
-        return self._call_sub_policies("remember", request, userid, **kw)
-
-    def forget(self, request):
-        """Get the correct headers to forget the current login."""
+    def identity(self, request) -> Identity | None:
+        return self._identity_cache.get_or_create(request)
 
+    def remember(self, request, userid, **kw):
         self._identity_cache.clear(request)
-
-        return self._call_sub_policies("forget", request)
-
-    def identity(self, request) -> Optional[Identity]:
-        """
-        Get an Identity object for valid credentials.
-
-        :param request: Pyramid request to inspect
-        """
-        return self._identity_cache.get_or_create(request)
+        return get_subpolicy(request).remember(request, userid, **kw)
 
     def _load_identity(self, request):
-        return self._call_sub_policies("identity", request)
-
-    def _call_sub_policies(self, method, request, *args, **kwargs):
-        """
-        Delegate calls to the correct set of security policies.
-
-        :param method: Method to call (like `identity()` or `forget()`)
-        :param request: Pyramid request object
-        :param args: Args to pass to the method
-        :param kwargs: Kwargs to pass to the method
-        :return: The response from the correct sub-policy
-        """
+        return get_subpolicy(request).identity(request)
 
-        if not is_api_request(request):
-            # This is usually the cookie policy for UI things
-            return getattr(self._ui_policy, method)(request, *args, **kwargs)
 
-        # Then we try the bearer header (or `access_token` GET param)
-        result = getattr(self._bearer_token_policy, method)(request, *args, **kwargs)
+@RequestLocalCache()
```

Comment:
Note the `@RequestLocalCache()` here so that `get_subpolicy()` will only be called once per request. This both saves CPU cycles and also makes sure that we don't delegate different method calls to different sub-policies within the same request.